paho/pinger.go

```diff
@@ -81,106 +73,67 @@ func (p *DefaultPinger) Run(conn net.Conn, keepAlive uint16) error {
 		return fmt.Errorf("conn is nil")
 	}
 	p.mu.Lock()
-	if p.timer != nil {
+	if p.running {
 		p.mu.Unlock()
-		return fmt.Errorf("Run() already called")
+		return fmt.Errorf("Run() already in progress")
 	}
-	select {
-	case <-p.done:
+	p.running = true
+	p.mu.Unlock()
+	defer func() {
+		p.mu.Lock()
+		p.running = false
 		p.mu.Unlock()
-		return fmt.Errorf("Run() called after stop()")
-	default:
+	}()
+
+	interval := time.Duration(keepAlive) * time.Second
+	timer := time.NewTimer(0) // Immediately send first pingreq
+	var lastPingSent time.Time
+	for {
+		select {
+		case <-ctx.Done():
+			timer.Stop() // We don't care if the timer has fired
+			return nil
+		case t := <-timer.C:
+			p.mu.Lock()
+			lastPingResponse := p.lastPingResponse
+			pingDue := p.lastPacketSent.Add(interval)
```

Comment:
"packet is sent" isn't equal to "packet is received". Are we sure that it always guarantees that the connection is not broken? Do we always receive something from the cloud in response (e.g. publishes with QoS 0)? Looks like it can significantly postpone the closure of connection in some edge cases.

Reply:
This is as per MQTT-3.1.2-20 (the only requirement to send a keep alive is if no control packet has been sent within the time frame). If you have a really broken TCP connection (that is something is ACKing packets that have not actually been delivered) then this could be an issue (but the MQTT protocol relies on a standards compliant connection).